packages/block-editor/src/components/block-list/moving-animation.js

```diff
@@ -0,0 +1,84 @@
+/**
+ * External dependencies
+ */
+import { useSpring, interpolate } from 'react-spring/web.cjs';
+
+/**
+ * WordPress dependencies
+ */
+import { useState, useLayoutEffect } from '@wordpress/element';
+import { useReducedMotion } from '@wordpress/compose';
+
+/**
+ * Hook used to compute the styles required to move a div into a new position.
+ *
+ * The way this animation works is the following:
+ *  - It first renders the element as if there was no animation.
+ *  - It takes a snapshot of the position of the block to use it
+ *    as a destination point for the animation.
+ *  - It restores the element to the previous position using a CSS transform
+ *  - It uses the "resetAnimation" flag to reset the animation
+ *    from the beginning in order to animate to the new destination point.
+ *
+ * @param {Object}  ref                      Reference to the element to animate.
+ * @param {boolean} isSelected               Whether it's the current block or not.
+ * @param {*}       triggerAnimationOnChange Variable used to trigger the animation if it changes.
+ *
+ * @return {Object} Style object.
+ */
+function useMovingAnimation( ref, isSelected, triggerAnimationOnChange ) {
+	const prefersReducedMotion = useReducedMotion();
```

Comment:
Curious: If we're eliminating the animation for reduced motion, can't it just return early as effectively `noop` ?

Reply:
It breaks the hooks rules.

Reply:
> It breaks the hooks rules.

Could you expand?

I'm familiar with the ["Don’t call Hooks inside [..] conditions" guideline](https://reactjs.org/docs/hooks-rules.html), but I'm not suggesting to use a hook inside the condition, rather to return early from the hook with an empty style object.

Reply:
It's the same though, Because there are hooks used after the early return. So for instance say `prefersReducedMotion` changes while the editor is loaded, it will break because React will confuse the next hook in the chain (in the previous render, it's the hook called after useMovingAnimation) with the next hook after the early return.

Reply:
> It's the same though, Because there are hooks used after the early return. So for instance say `prefersReducedMotion` changes while the editor is loaded, it will break because React will confuse the next hook in the chain (in the previous render, it's the hook called after useMovingAnimation) with the next hook after the early return.

Oh, I see now, it's obvious in retrospect; an early return here is just another way of expressing the inverse condition of the remainder of the function (with hooks) being in the condition `if ( ! prefersReducedMotion ) {`